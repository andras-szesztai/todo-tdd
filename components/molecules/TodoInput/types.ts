export interface Props {
    onSubmit: (newTodo: string) => void
}
