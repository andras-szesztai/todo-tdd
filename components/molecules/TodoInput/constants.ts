export const TODO_INPUT_LABEL = 'Todo Title'
export const TODO_INPUT_PLACEHOLDER = 'Enter a Todo title...'
export const TODO_INPUT_BUTTON_TEXT = 'Submit'

export const TODO_INPUT_ID = 'todo-input'
