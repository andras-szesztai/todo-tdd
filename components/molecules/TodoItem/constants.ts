export const TODO_COMPLETE_BUTTON_TEXT = {
    complete: 'Complete',
    unComplete: 'Un-complete',
}
