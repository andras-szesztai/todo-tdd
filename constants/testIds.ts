export const TEST_ID = {
    todoInputLabel: 'todo-input-label',
    todoInput: 'todo-input',
    todoInputButton: 'todo-input-button',
    todoViewContainer: 'todos-view-container',
    todoListContainer: 'todo-list-container',
    todoItem: 'todo-item',
    todoCompleteButton: 'todo-complete-button',
}
