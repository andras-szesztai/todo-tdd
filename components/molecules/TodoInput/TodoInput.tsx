'use client'

import { useState } from 'react'

import { Button } from 'components/atoms/Button'
import { Input } from 'components/atoms/Input'
import { Label } from 'components/atoms/Label'
import { TEST_ID } from 'constants/testIds'

import {
    TODO_INPUT_BUTTON_TEXT,
    TODO_INPUT_LABEL,
    TODO_INPUT_PLACEHOLDER,
    TODO_INPUT_ID,
} from './constants'
import { Props } from './types'

function TodoInput({ onSubmit }: Props) {
    const [value, setValue] = useState('')
    return (
        <div className="flex items-end gap-2">
            <div className="flex flex-col gap-2">
                <div className="ml-1">
                    <Label
                        label={TODO_INPUT_LABEL}
                        testId={TEST_ID.todoInputLabel}
                        htmlFor={TODO_INPUT_ID}
                    />
                </div>
                <Input
                    onChange={(value) => {
                        setValue(value)
                    }}
                    value={value}
                    placeholder={TODO_INPUT_PLACEHOLDER}
                    id={TODO_INPUT_ID}
                />
            </div>
            <Button
                text={TODO_INPUT_BUTTON_TEXT}
                disabled={!value}
                onClick={() => {
                    onSubmit(value)
                    setValue('')
                }}
            />
        </div>
    )
}

export default TodoInput
