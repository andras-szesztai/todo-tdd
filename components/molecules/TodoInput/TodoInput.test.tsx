import { fireEvent, render } from '@testing-library/react'

import { TEST_ID } from 'constants/testIds'

import TodoInput from './TodoInput'
import {
    TODO_INPUT_BUTTON_TEXT,
    TODO_INPUT_LABEL,
    TODO_INPUT_PLACEHOLDER,
} from './constants'

describe('TodoInput', () => {
    it('renders default state with correct label, without any value and with correct placeholder in input and submit button disabled', () => {
        const { getByTestId, asFragment } = render(<TodoInput />)
        const label = getByTestId(TEST_ID.todoInputLabel)
        expect(label).toBeInTheDocument()
        expect(label.textContent).toBe(TODO_INPUT_LABEL)
        const input = getByTestId(TEST_ID.todoInput)
        expect(input).toBeInTheDocument()
        expect(input).toHaveAttribute('placeholder', TODO_INPUT_PLACEHOLDER)
        expect(input).not.toHaveFocus()
        const button = getByTestId(TEST_ID.todoInputButton)
        expect(button).toBeInTheDocument()
        expect(button.textContent).toBe(TODO_INPUT_BUTTON_TEXT)
        expect(button).toBeDisabled()
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders input field that can be type into resulting in enabled submit button', () => {
        const nextValue = 'test todo'
        const { getByTestId, asFragment } = render(<TodoInput />)
        const input = getByTestId(TEST_ID.todoInput)
        const button = getByTestId(TEST_ID.todoInputButton)
        expect(input).toHaveValue('')
        expect(button).toBeDisabled()
        fireEvent.change(input, { target: { value: nextValue } })
        expect(input).toHaveValue(nextValue)
        expect(button).not.toBeDisabled()
        expect(asFragment()).toMatchSnapshot()
    })
})
