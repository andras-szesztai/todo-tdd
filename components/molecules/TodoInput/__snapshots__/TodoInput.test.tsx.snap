// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoInput renders default state with correct label, without any value and with correct placeholder in input and submit button disabled 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 items-end"
  >
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="ml-1"
      >
        <label
          class="font-semibold text-xl text-primary-dark cursor-pointer select-none"
          data-testid="todo-input-label"
          for="todo-input"
        >
          Todo Title
        </label>
      </div>
      <input
        class="w-60 rounded-lg border font-medium border-primary-dark focus:outline-none focus:border-accent text-base p-2 placeholder-primary-dark-disabled text-primary-dark"
        data-testid="todo-input"
        id="todo-input"
        placeholder="Enter a Todo title..."
        value=""
      />
    </div>
    <button
      class="inline font-medium p-2 rounded-lg text-white bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent"
      data-testid="todo-input-button"
      disabled=""
      type="button"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;

exports[`TodoInput renders input field that can be type into resulting in enabled submit button 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 items-end"
  >
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="ml-1"
      >
        <label
          class="font-semibold text-xl text-primary-dark cursor-pointer select-none"
          data-testid="todo-input-label"
          for="todo-input"
        >
          Todo Title
        </label>
      </div>
      <input
        class="w-60 rounded-lg border font-medium border-primary-dark focus:outline-none focus:border-accent text-base p-2 placeholder-primary-dark-disabled text-primary-dark"
        data-testid="todo-input"
        id="todo-input"
        placeholder="Enter a Todo title..."
        value="test todo"
      />
    </div>
    <button
      class="inline font-medium p-2 rounded-lg text-white bg-primary-dark border cursor-pointer select-none border-primary-dark disabled:bg-primary-dark-disabled disabled:cursor-not-allowed focus:outline-none focus:border-accent"
      data-testid="todo-input-button"
      type="button"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;
